src/core/components/Experiment/index.js

```diff
@@ -0,0 +1,158 @@
+/* @flow */
+import * as React from 'react';
+import invariant from 'invariant';
+import cookie from 'react-cookie';
+import { compose } from 'redux';
+import { connect } from 'react-redux';
+
+import withUIState from 'core/withUIState';
+import tracking from 'core/tracking';
+import type { AppState } from 'amo/store';
+
+type Props = {|
+  AName: string,
+  BName: string,
+  nameId: string,
+|};
+
+type UIStateType = {|
+  abTest: string | void,
+|};
+
+type InternalProps = {|
+  ...Props,
+  _cookie: typeof cookie,
+  _tracking: typeof tracking,
+  randomizer: () => number,
+  setUIState: ($Shape<UIStateType>) => void,
+  uiState: UIStateType,
+  WrappedComponent: Function,
+|};
+
+const initialUIState: UIStateType = {
+  abTest: undefined,
+};
+
+export function mergeExperimentProps({
+  WrappedComponent,
+  nameId,
+  AName,
+  BName,
+}: {|
+  WrappedComponent: Function,
+  nameId: string,
+  AName: string,
+  BName: string,
+|}) {
+  return function mapStateToProps(state: AppState, ownProps: Object) {
+    return {
+      ...ownProps,
+      nameId,
+      AName,
+      BName,
+      WrappedComponent,
+    };
+  };
+}
+
+export class Experiment extends React.Component<InternalProps> {
+  abTestCookie: string | void;
+
+  static defaultProps = {
+    _cookie: cookie,
+    _tracking: tracking,
+    randomizer: Math.random,
+  };
+
+  componentDidMount() {
+    const { _cookie, _tracking, AName, BName, nameId, randomizer } = this.props;
+
+    this.abTestCookie = _cookie.load(`AB_${nameId}_COOKIE`);
+
+    if (this.abTestCookie === undefined) {
+      this.abTestCookie =
+        randomizer() >= 0.5 ? `AB_${nameId}_${AName}` : `AB_${nameId}_${BName}`;
+      _cookie.save(`AB_${nameId}_COOKIE`, this.abTestCookie, {
+        path: '/', // TODO: make this flexible too possibly.
+      });
+    }
+
+    if (this.abTestCookie) {
+      this.props.setUIState({
+        abTest: this.abTestCookie,
+      });
+
+      _tracking.sendEvent({
+        action: `${nameId} Page View`,
+        category: this.abTestCookie ? `AMO ${this.abTestCookie}` : '',
+        label: '',
+      });
+    }
+  }
+
+  trackClick = (e: SyntheticEvent<any>, url: string = '') => {
+    const { _cookie, _tracking, nameId } = this.props;
+
+    const nodeType = e.currentTarget.nodeName.toLowerCase();
+
+    if (nodeType === 'a') {
+      _tracking.sendEvent({
+        action: `${nameId} Click`,
+        category: `AMO ${_cookie.load(`AB_${nameId}_COOKIE`)}`,
+        label: url,
+      });
+    }
+  };
+
+  render() {
+    const { WrappedComponent, _cookie, nameId, uiState, ...props } = this.props;
+
+    invariant(nameId, 'nameId is required');
+
+    const exposedPropHelpers = {
+      trackClick: (...args) => this.trackClick(...args),
+    };
+
+    // When the cookie is initially set it will not be seen on client
+    // until there is a refresh so we are also checking uiState.
```

Comment:
`componentDidMount` is only executed on the client, so we will never set a cookie on the server if I understand correctly. Do you really need this? That was not present before AFAIK and that would simplify lots of things here.

Reply:
I tried doing this in componentWillMount but I thought I saw a complaint about the classes not matching (on server and client). I can move it back/try that again tho...

To be clear tho - do I really need what? :) the uiState?

Reply:
hmm I guess I don't see this now (about the class mismatch so I will it back and then this will resolve - i believe - the client issue noted above)